### PR TITLE
Updates to the historicalv2 endpoint.

### PR DESCRIPTION
**Pull request by KevinJoiner**
# Proposed Changes
- Updates to the historicalv2 endpoint to support start and end times. up to 1ms resolution.
- Updates to the historicalv2 endpoint to support adaptive bucket sizes.
- Updates the permissions for the historicalv2 endpoint. Only users with AllTimeLocation permission can access historical location data.
- Moves database-related code to the services package.
- Adds a test package for an elastic container with test data.
### Impacted Routes

This change will change `/v2/vehicle/:tokenID/history` endpoint

### CODE REVIEW

File: internal/services/elastic/elastic.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+// EsClient returns the underlying elastic client used by the service.
+//
+// This function exists to maintain previous behavior for endpoint who are not yet using the elastic service.
+func (s *Service) EsClient() *elasticsearch.TypedClient {
+	return s.esClient
+}
+
+// GetHistoryParams defines the parameters for retrieving history from elastic.
+type GetHistoryParams struct {
+	// StartTime is the start time of the history.
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:68]: Could we mention the defaults behavior somewhere? Maybe here?

File: internal/controllers/historicalv2.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	startTime, err := getTimeFromQuery(c, "startTime")
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, err.Error())
+	}
+	endTime, err := getTimeFromQuery(c, "endTime")
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, err.Error())
+	}
+
+	buckets, err := strconv.ParseInt(c.Query("buckets", "0"), 10, 64)
```
Comment (elffjs) [inline internal/controllers/historicalv2.go:37]: Do we want to cap this? Couldn't it get out of hand?

File: internal/services/elastic/elastic.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+			},
+		},
+		Size: ptr(0), // only return the aggregations
+	}
+}
+
+// calculateInterval calculates the desired interval in milliseconds. This interval will provide n buckets between the start and end times.
+// if the interval is 0 or negative it will be clamped to 1ms.
+func calculateInterval(start, end time.Time, buckets int) int {
+	duration := int(end.Sub(start).Milliseconds())
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:190]: Do we check that `start` and `end` are "in order" anywhere?

Comment (elffjs) [reply internal/services/elastic/elastic.go:190]: Note to myself: `int` is 64 bits on 64 bit systems, so this cast is okay.

File: internal/controllers/historicalv2.go (hunk 2 above)
Comment (elffjs) [reply internal/controllers/historicalv2.go:37]: What if it's negative?

File: internal/controllers/historicalv2.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+// @Param        tokenID  path   int64  true   "token id"
+// @Param        startTime     query  string  false  "startTime is an RFC3339 formatted date-time string. if empty two weeks back"
+// @Param        endTime       query  string  false  "endTime is an RFC3339 formatted date-time string. if empty today"
+// @Param        buckets       query  string  false  "number of data points to return, default 1000"
+// @Security     BearerAuth
+// @Router       /v2/vehicle/{tokenID}/history [get]
+func (d *DeviceDataController) GetHistoricalPermissionedV2(c *fiber.Ctx) error {
+	// get the startTime and endTime from the query parameters and convert them to time.Time
+	var err error
+	startTime, err := getTimeFromQuery(c, "startTime")
```
Comment (elffjs) [inline internal/controllers/historicalv2.go:28]: Did you see [`QueryParser`](https://docs.gofiber.io/api/ctx/#queryparser)? Does that help at all?

File: internal/controllers/historicalv2.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+		EndTime:      endTime,
+		Buckets:      int(buckets),
+		PrivilegeIDs: claims.PrivilegeIDs,
+	}
+
+	result, err := d.esService.GetHistory(ctx, params)
+	if err != nil {
+		return fmt.Errorf("failed to get history: %w", err)
+	}
+	resp := historyResp{Status: result}
```
Comment (elffjs) [inline internal/controllers/historicalv2.go:73]: Hum, was it always `"status"` singular? That seems off.

File: internal/constants/consts.go [hunk 6]
```diff
@@ -37,3 +37,11 @@ const (
 	UserDeviceAPIIntegrationStatusDuplicateIntegration  string = "DuplicateIntegration"
 	UserDeviceAPIIntegrationStatusAuthenticationFailure string = "AuthenticationFailure"
 )
+
+// PrivilegeIDs for claims.
+const (
+	NonLocationData int64 = 1
```
Comment (elffjs) [inline internal/constants/consts.go:43]: Maybe we should pull these out to the shared repo.

File: internal/services/elastic/elastic.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+	// build the query
+	req := buildHistoryQuery(params)
+
+	// log requests
+	reqb, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request: %w", err)
+	}
+	s.log.Info().RawJSON("request", reqb).Str("index", s.settings.DeviceDataIndexNameV2).Msg("V2 request")
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:95]: This could get very spammy, but it's good for now.

File: internal/services/elastic/elastic.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+	// perform the request
+	res, err := s.esClient.Search().Index(s.settings.DeviceDataIndexNameV2).Request(&req).Perform(ctx)
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:98]: Help me—what's the difference between `Do` and `Perform`?

File: internal/services/elastic/elastic.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+// buildHistoryQuery builds the query to retrieve the history of a device from elastic.
+func buildHistoryQuery(params GetHistoryParams) search.Request {
+	// convert the start and end times to milliseconds since epoch
+	startArg := strconv.Itoa(int(params.StartTime.UnixMilli()))
+	endArg := strconv.Itoa(int(params.EndTime.UnixMilli()))
+
+	// calculate the interval in milliseconds
+	interval := calculateInterval(params.StartTime, params.EndTime, params.Buckets)
+	intervalArg := strconv.Itoa(interval) + "ms"
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:144]: I'm a little worried about how big their types are here. Already a month is too large for a Java `int`.

File: internal/services/elastic/elastic.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+// getHitsFromHistoryResponse extracts the hits from the response body.
+// The maxValues parameter is used to limit the number of values returned.
+func getHitsFromHistoryResponse(body []byte, maxValues int) []any {
+	retData := []any{}
+	bucketKey := fmt.Sprintf("aggregations.%s.buckets", docByInterval)
+	hitsKey := fmt.Sprintf("%s.hits.hits", singleDoc)
+	gjson.GetBytes(body, bucketKey).ForEach(func(key, v gjson.Result) bool {
+		gjson.Get(v.Raw, hitsKey).ForEach(func(key, value gjson.Result) bool {
+			// if we have reached the max number of values, stop
+			if len(retData) == maxValues {
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:207]: How would we go above this?

File: internal/services/elastic/elastic.go [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+// getHitsFromHistoryResponse extracts the hits from the response body.
+// The maxValues parameter is used to limit the number of values returned.
+func getHitsFromHistoryResponse(body []byte, maxValues int) []any {
+	retData := []any{}
+	bucketKey := fmt.Sprintf("aggregations.%s.buckets", docByInterval)
+	hitsKey := fmt.Sprintf("%s.hits.hits", singleDoc)
+	gjson.GetBytes(body, bucketKey).ForEach(func(key, v gjson.Result) bool {
+		gjson.Get(v.Raw, hitsKey).ForEach(func(key, value gjson.Result) bool {
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:205]: @Allyson-English I assume you wrote this logic—what about this?

https://go.dev/play/p/jlOEauUPOfn

File: internal/services/elastic/elastic.go [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+	}
+	eClient := Service{
+		esClient: es8Client,
+		log:      logger,
+		settings: settings,
+	}
+	return &eClient, nil
+}
+
+// EsClient returns the underlying elastic client used by the service.
```
[the remark this answers is not in the document]
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:59]: Yes

File: internal/services/elastic/elastic.go (hunk 11 above)
Comment (elffjs) [reply internal/services/elastic/elastic.go:205]: But do I want to do this at all? I feel like we're parsing this thing when we do `.Value()`. Can we `json.RawMessage` this somehow?

File: internal/services/elastic/elastic.go (hunk 1 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:68]: Can do. On that note, Do you think the default behavior should be an explicit method of this structure? 
`params.SetDefaults()` that way the caller can choose if they want to use them or not.

File: .gitignore [hunk 13]
```diff
@@ -14,4 +14,8 @@
 # Dependency directories (remove the comment below to include it)
 # vendor/
 settings.yaml
-.idea/
\ No newline at end of file
+.idea/
+
+# Igonre the device-data-api binary only
+device-data-api
+!device-data-api/
```
[the remark this answers is not in the document]
Comment (KevinJoiner) [reply .gitignore:21]: I almost committed the binary so I figured it doesn't hurt to add here.

File: internal/services/elastic/elastic.go (hunk 7 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:95]: I left it here because it was part of the original implementation, but I do not love that we have to marshal the query just to log it. If you are okay with it, I would prefer to remove this log.

File: internal/services/elastic/elastic.go (hunk 9 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:144]: I will add a test case for this and see how it is handled.

File: internal/services/elastic/elastic.go (hunk 3 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:190]: No, but we do test that this does not break anything. If this is a negative number, the interval will be negative and be clamped to 1. Elastic will return an empty list since no time will match your time range.

File: internal/services/elastic/elastic.go (hunk 11 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:205]: We should be able to use RAW since this output is eventually converted back to JSON when it's sent in the response.

File: internal/services/elastic/elastic.go (hunk 1 above)
Comment (elffjs) [reply internal/services/elastic/elastic.go:68]: That's a good question.

File: internal/services/elastic/elastic.go (hunk 10 above)
Comment (elffjs) [reply internal/services/elastic/elastic.go:207]: Ah, when we round down when we calculate the interval.

Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:207]: It has to do with the exact time your start and end fall on and how elastic does rounding for it's aggregation.

If you look through the test cases I added one called "buckets + 1 returned from elastic" and you can see that those exact start and stop time will cause elastic to provide an extra bucket.

Formula and docs on how bucket key is determined.
[Docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-datehistogram-aggregation.html)
> bucket_key = Math.floor(time / interval) * interval

File: internal/constants/consts.go (hunk 6 above)
Comment (KevinJoiner) [reply internal/constants/consts.go:43]: 100% agree since they are also used in device-data.

File: internal/services/elastic/elastic.go (hunk 3 above)
Comment (elffjs) [reply internal/services/elastic/elastic.go:190]: > Do we check that `start` and `end` are "in order" anywhere?

Ah, there's a negative check.

File: internal/controllers/historicalv2.go (hunk 4 above)
Comment (KevinJoiner) [reply internal/controllers/historicalv2.go:28]: I saw it, but I would still need to convert the fields to `startTime` and `endTime` to `time.Time` objects and `buckes` to an int so I didn't think the abstraction had a huge benefit. I am not against it if we want to start using it consistently.

File: internal/services/elastic/elastic.go [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+				},
+			},
+		},
+		Size: ptr(0), // only return the aggregations
+	}
+}
+
+// calculateInterval calculates the desired interval in milliseconds. This interval will provide n buckets between the start and end times.
+// if the interval is 0 or negative it will be clamped to 1ms.
+func calculateInterval(start, end time.Time, buckets int) int {
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:189]: Maybe put  milliseconds or ms in the name?

File: internal/controllers/historicalv2.go (hunk 5 above)
Comment (KevinJoiner) [reply internal/controllers/historicalv2.go:73]: It was before, but its v2 so now would be the time to update it https://github.com/DIMO-Network/device-data-api/blob/473e4673e2db8c27b160c6a35d466b3634909a28/internal/controllers/device_data_controller.go#L396

File: internal/services/elastic/elastic.go (hunk 8 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:98]: Do to the import of the elastic testcontainer I had to bump the elastic version from `v8.6.0 => v8.11.1` In this bump they updated the package so `Do()` no longer returns a `http.Response` and now returns an `typed.Response` and `Perform()` returns `http.Response` so to maintain previous behavior I had to change the calls to `Perform`

https://github.com/elastic/go-elasticsearch/releases/tag/v8.7.0
> Do method on endpoints now return a typed response, one per endpoint.
Perform method added on endpoints, returns http.Response as did Do.

File: internal/controllers/historicalv2.go (hunk 4 above)
Comment (elffjs) [reply internal/controllers/historicalv2.go:28]: I think that function would do the conversion for you. That's what I would be interested in.

File: internal/controllers/historicalv2.go (hunk 5 above)
Comment (elffjs) [reply internal/controllers/historicalv2.go:73]: Let's change it now.

File: internal/services/elastic/elastic.go (hunk 7 above)
Comment (elffjs) [reply internal/services/elastic/elastic.go:95]: Raise it to Debug level maybe?

File: internal/services/elastic/elastic.go (hunk 10 above)
Comment (elffjs) [reply internal/services/elastic/elastic.go:207]: That makes sense to me. Maybe leave a comment? I know tests are documentation and blah blah.

File: internal/controllers/historicalv2.go (hunk 2 above)
Comment (KevinJoiner) [reply internal/controllers/historicalv2.go:37]: > Do we want to cap this? Couldn't it get out of hand?

The natural cap is the number of ms between start and stop time. Our other cap would be the context, so 5 minutes, which is probably too long. We can either give it a shorter timeout or an arbitrary size cap.

> What if it's negative?

Negative cause weird behavior with the interval calculation I am going to treat `buckets <= 0` as invalid.

File: internal/services/elastic/elastic.go (hunk 8 above)
Comment (elffjs) [reply internal/services/elastic/elastic.go:98]: Got it. I wonder which is better.

File: internal/services/elastic/elastic.go [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+		return nil, fmt.Errorf("failed to complete request: %w", err)
+	}
+
+	//nolint:errcheck // we don't care about the error closing the body.
+	defer res.Body.Close()
+
+	// check for errors any status code over 299 is an error
+	if res.StatusCode >= http.StatusMultipleChoices {
+		body, _ := io.ReadAll(res.Body) // we ignore the error here because we are already in an error state
+		return nil, fmt.Errorf("got status code %d from Elastic message = %s.", res.StatusCode, string(body))
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:109]: Do you agree with the linter?

File: internal/services/elastic/elastic.go (hunk 3 above)
Comment (elffjs) [reply internal/services/elastic/elastic.go:190]: This is one of a few places where it's kinda hard to follow. It's okay though, there are tests.

File: internal/services/elastic/elastic_test.go [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+				for _, field := range tc.expectedFields {
+					fieldData := gjson.GetBytes(hitData, field)
+					require.Truef(t, fieldData.Exists(), "expected field %q not found in result", field)
+				}
+				for _, field := range tc.excludedFields {
+					fieldData := gjson.GetBytes(hitData, field)
+					require.Falsef(t, fieldData.Exists(), "excluded field %q found in result", field)
+				}
+			}
+			slices.Sort(tc.expectedIds)
```
Comment (elffjs) [inline internal/services/elastic/elastic_test.go:266]: We do kinda care about the order though, yeah?

File: internal/services/elastic/elastic.go (hunk 11 above)
Comment (Allyson-English) [reply internal/services/elastic/elastic.go:205]: I think gjson is a relic of wanted to maniuplate the data instead of just passing it through... are you thinking something like this? https://go.dev/play/p/zOle8BP1YNy

File: internal/services/elastic/elastic.go (hunk 1 above)
Comment (Allyson-English) [reply internal/services/elastic/elastic.go:68]: I like this idea! I believe the params used in the history endpoint are used elsewhere.. do we want to have a more generic params struct? (realizing the defaults may be different elsewhere though so I can see the case for having a specific history one)

Comment (elffjs) [reply internal/services/elastic/elastic.go:68]: Where else?

Comment (Allyson-English) [reply internal/services/elastic/elastic.go:68]: ah, nvm-- I wasnt looking at this closely enough I was thinking of http params elsewhere

File: internal/services/elastic/elastic.go (hunk 11 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:205]: Here are two versions of this 
1. Uses the typed Response object from elastic 
2. Uses gjson without allocations similar to what @Allyson-English provided

The gjson should be a bit performant since we don't have to unmarshal from elastic then remarshal for sending the response.
``` golang
// getHitsFromHistoryResponse extracts the hits from the response body.
// The maxValues parameter is used to limit the number of values returned.
func getHitsFromHistoryResponse(body []byte, maxValues int) []json.RawMessage {
	retData := []json.RawMessage{}
	// This key gets the first data point in each bucket which should only contain 1 value
	fullKey := fmt.Sprintf("aggregations.%s.buckets.#.%s.hits.hits.0._source", docByInterval, singleDoc)
	gjson.GetBytes(body, fullKey).ForEach(func(key, value gjson.Result) bool {
		// if we have reached the max number of values, stop
		if len(retData) == maxValues {
			return false
		}
		retData = append(retData, resultToBytes(body, value))
		return true
	})
	return retData
}

// resultToBytes converts the result to a byte slice.
// this logic is the recommended way to extract the raw bytes from a gjson result.
// more info: https://github.com/tidwall/gjson/blob/6ee9f877d683381343bc998c137339c7ae908b86/README.md#working-with-bytes
func resultToBytes(body []byte, result gjson.Result) []byte {
	var raw []byte
	if result.Index > 0 {
		raw = body[result.Index : result.Index+len(result.Raw)]
	} else {
		raw = []byte(result.Raw)
	}
	return raw
}

// getHitsFromHistoryResponseTyped extracts the hits from the response body.
func getHitsFromHistoryResponseTyped(resp *search.Response, maxValues int) ([]json.RawMessage, error) {
	retData := []json.RawMessage{}
	datAgg, ok := resp.Aggregations[docByInterval].(*types.DateHistogramAggregate)
	if !ok {
		return retData, fmt.Errorf("failed to get data from response")
	}
	buckets, ok := datAgg.Buckets.([]types.DateHistogramBucket)
	if !ok {
		return retData, fmt.Errorf("failed to get data from response")
	}
	for _, bucket := range buckets {
		singAgg, ok := bucket.Aggregations[singleDoc].(*types.TopHitsAggregate)
		if !ok {
			return retData, fmt.Errorf("failed to get data from response")
		}
		for _, hit := range singAgg.Hits.Hits {
			if len(retData) >= maxValues {
				return retData, nil
			}
			retData = append(retData, hit.Source_)
		}
	}

	return retData, nil
}

```

File: internal/services/elastic/elastic.go (hunk 15 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:109]: Yeah the period was an accident.

File: internal/services/elastic/elastic_test.go (hunk 16 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic_test.go:266]: Since we do not use the [BucketSort](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-pipeline-bucket-sort-aggregation.html#search-aggregations-pipeline-bucket-sort-aggregation) Aggregation we can't guarantee the order. Though I do think it would be beneficial to sort by time.

Comment (elffjs) [reply internal/services/elastic/elastic_test.go:266]: From [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-datehistogram-aggregation.html#date-histogram-order),

> By default the returned buckets are sorted by their key ascending, but you can control the order using the order setting. This setting supports the same order functionality as [Terms Aggregation](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#search-aggregations-bucket-terms-aggregation-order).

File: internal/controllers/device_data_controller.go [hunk 17]
```diff
@@ -1,72 +1,74 @@
...
+const (
+	defaultTimeout = time.Minute * 5
 )
 
 type DeviceDataController struct {
 	Settings        *config.Settings
 	log             *zerolog.Logger
 	deviceAPI       services.DeviceAPIService
-	es8Client       *es8.TypedClient
+	esService       EsInterface
```
Comment (elffjs) [inline internal/controllers/device_data_controller.go:49]: Oh weeu I didn't notice this name this is Enterprise naming.

File: internal/controllers/historicalv2.go (hunk 2 above)
Comment (elffjs) [reply internal/controllers/historicalv2.go:37]: Letting the timeout cap it is a bit spooky to me but I'm fine to leave it.

File: internal/services/elastic/elastic.go (hunk 11 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:205]: `❤️ == gjson` All the casting when using the typed response hurts.

File: internal/controllers/historicalv2_test.go [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+			app := fiber.New()
+			app.Use(test.ClaimsInjectorTestHandler(tc.customClaims))
+			app.Get("/v2/vehicle/:tokenID/history", deviceCtrl.GetHistoricalPermissionedV2)
+
+			// if we do not expect errors, we need to set the expected calls
+			if tc.expectedResponse == fiber.StatusOK {
+				esMock.EXPECT().GetHistory(gomock.Any(), gomock.AssignableToTypeOf(elastic.GetHistoryParams{})).DoAndReturn(func(ctx context.Context, params elastic.GetHistoryParams) (interface{}, error) {
+					require.Equal(t, tc.expectedReq, params)
+					return []any{}, nil
+				})
```
Comment (elffjs) [inline internal/controllers/historicalv2_test.go:161]: Do we gain something here over doing

```
esMock.EXPECT().GetHistory(gomock.Any(), tc.expectedReq).Return([]any{}, nil)
```

File: internal/services/elastic/elastic_test.go [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+			slices.Sort(ids)
+			require.Truef(t, slices.Equal(tc.expectedIds, ids), "expected ids: %v, got ids: %v", tc.expectedIds, ids)
+		})
+	}
+}
+
+// setupService creates a new elastic container and a new elastic service with the client and the logger.
+func setupService(t *testing.T, ctx context.Context) (*elastic.Service, func()) {
+	t.Helper()
+	settings, cleanup, err := elasticcontainer.Create(ctx)
```
Comment (elffjs) [inline internal/services/elastic/elastic_test.go:276]: I forgot they had a module just for this. This is nice.

File: internal/services/elastic/elastic.go (hunk 1 above)
Comment (elffjs) [reply internal/services/elastic/elastic.go:68]: A method seems cool.

Comment (elffjs) [reply internal/services/elastic/elastic.go:71]: It sounds like we are not doing this.

File: internal/services/elastic/elastic.go (hunk 9 above)
Comment (elffjs) [reply internal/services/elastic/elastic.go:143]: It worked out?

File: internal/services/elastic/elastic.go [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+	if result.Index > 0 {
+		raw = body[result.Index : result.Index+len(result.Raw)]
+	} else {
+		raw = []byte(result.Raw)
+	}
+	return raw
+}
+
+// getHistoricalPrivilegeFilter returns the location filter based on the privileges.
+func getHistoricalPrivilegeFilter(privs []privileges.Privilege) types.SourceFilter {
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:230]: This reads well now.

File: internal/services/elastic/elastic.go [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+	"github.com/DIMO-Network/shared/privileges"
+	"github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/typedapi/core/search"
+	"github.com/elastic/go-elasticsearch/v8/typedapi/types"
+	"github.com/rs/zerolog"
+	"github.com/tidwall/gjson"
+)
+
+const (
+	docByInterval  = "documents_by_interval"
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:24]: This is a nice touch.

File: internal/services/elastic/elastic.go [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+		return 1 // 0 or negative interval is not supported, clamp to 1
+	}
+	return interval
+}
+
+// getHitsFromHistoryResponse extracts the hits from the response body.
+// The maxValues parameter is used to limit the number of values returned.
+// Due to the mapping function used by elaticsearch `bucket_key = Math.floor(time / interval) * interval
+// the number of buckets returned may be 1 more than the number of buckets requested.
+func getHitsFromHistoryResponse(body []byte, maxValues int) []json.RawMessage {
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:201]: Good find.

File: internal/services/elastic/elastic.go [hunk 23]
```diff
@@ -0,0 +1,257 @@
...
+		retData = append(retData, resultToBytes(body, value))
+		return true
+	})
+	return retData
+}
+
+// resultToBytes converts the result to a byte slice.
+// this logic is the recommended way to extract the raw bytes from a gjson result.
+// more info: https://github.com/tidwall/gjson/blob/6ee9f877d683381343bc998c137339c7ae908b86/README.md#working-with-bytes
+func resultToBytes(body []byte, result gjson.Result) []byte {
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:219]: Good find.

File: internal/services/elastic/elastic.go [hunk 24]
```diff
@@ -0,0 +1,257 @@
...
+	return interval
+}
+
+// getHitsFromHistoryResponse extracts the hits from the response body.
+// The maxValues parameter is used to limit the number of values returned.
+// Due to the mapping function used by elaticsearch `bucket_key = Math.floor(time / interval) * interval
+// the number of buckets returned may be 1 more than the number of buckets requested.
+func getHitsFromHistoryResponse(body []byte, maxValues int) []json.RawMessage {
+	retData := []json.RawMessage{}
+	// This key gets the first data point in each bucket which should only contain 1 value
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:203]: At most one value, I guess.

File: internal/services/elastic/elastic.go [hunk 25]
```diff
@@ -0,0 +1,257 @@
...
+	// depending on the start and end time we may get len(results) <=  buckets + 1
+	// pass the max number of values to getHitsFromHistoryResponse to ensure we only get the desired number of values
+	return getHitsFromHistoryResponse(body, params.Buckets), nil
+}
+
+// SetDefaultHistoryParams sets the default values for the history params.
+// If the buckets are less than or equal to 0, it will be set to 1000.
+// If the end time is zero, it will be set to the current time.
+// If the start time is zero, it will be set to 2 weeks from the end time.
+func (g *GetHistoryParams) SetDefaultHistoryParams() {
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:122]: Oh, we are doing this. Could we put it next to the struct?

File: internal/test/elasticcontainer/elasticcontainer.go [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+	}
+	if err != nil {
+		return esmodule.Options{}, cleanup, fmt.Errorf("Could not start Elasticsearch container: %v", err)
+	}
+
+	return esContainer.Settings, cleanup, nil
+}
+
+// AddVSSMapping inserts a mapping into Elasticsearch using a typed client.
+func AddVSSMapping(ctx context.Context, client *elasticsearch.TypedClient, index string) error {
```
Comment (elffjs) [inline internal/test/elasticcontainer/elasticcontainer.go:39]: I'm worried that this is just inserting the mapping as a document.

File: internal/controllers/gomock_reflect_3249397197/prog.go [hunk 27]
```diff
@@ -0,0 +1,67 @@
+
```
Comment (elffjs) [inline internal/controllers/gomock_reflect_3249397197/prog.go:1]: Delete?

**Comment by elffjs** [comment]
Should we regenerate the Swagger? [The command is in the README.](https://github.com/DIMO-Network/device-data-api?tab=readme-ov-file#swagger-generation)

File: cmd/device-data-api/api.go [hunk 28]
```diff
@@ -100,12 +94,12 @@ func startWebAPI(logger zerolog.Logger, settings *config.Settings, dbs func() *d
 	vehicleAddr := common.HexToAddress(settings.VehicleNFTAddress)
 
 	// token based routes
-	vTokenV1.Get("/history", tk.OneOf(vehicleAddr, []int64{controllers.NonLocationData, controllers.AllTimeLocation}), cacheHandler, deviceDataController.GetHistoricalRawPermissioned)
-	vTokenV1.Get("/status", tk.OneOf(vehicleAddr, []int64{controllers.NonLocationData, controllers.CurrentLocation, controllers.AllTimeLocation}), cacheHandler, deviceDataController.GetVehicleStatus)
-	vTokenV1.Get("/status-raw", tk.OneOf(vehicleAddr, []int64{controllers.NonLocationData, controllers.CurrentLocation, controllers.AllTimeLocation}), cacheHandler, deviceDataController.GetVehicleStatusRaw)
+	vTokenV1.Get("/history", tk.OneOf(vehicleAddr, []int64{int64(privileges.NonLocationData), int64(privileges.AllTimeLocation)}), cacheHandler, deviceDataController.GetHistoricalRawPermissioned)
```
Comment (elffjs) [inline cmd/device-data-api/api.go:97]: Maybe it should've been `int64` in shared? Whoops.

File: internal/services/elastic/elastic.go [hunk 29]
```diff
@@ -0,0 +1,257 @@
...
+	return search.Request{
+		// get documents that match the device ID, are within the time range, and have a vehicle field
+		Query: &types.Query{
+			Bool: &types.BoolQuery{
+				Filter: []types.Query{
+					{Term: map[string]types.TermQuery{subjectField: {Value: params.DeviceID}}},
+					{Range: map[string]types.RangeQuery{timeField: types.DateRangeQuery{Gte: &startArg, Lte: &endArg}}},
+				},
+				Should: []types.Query{
+					{Exists: &types.ExistsQuery{Field: "data.vehicle"}},
```
Comment (elffjs) [inline internal/services/elastic/elastic.go:157]: @Allyson-English Is there a reason this `exists` thing is here in should and not in the filter?

File: internal/services/elastic/elastic.go (hunk 25 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:122]: Yeah

File: cmd/device-data-api/api.go (hunk 28 above)
Comment (KevinJoiner) [reply cmd/device-data-api/api.go:97]: Opened a PR if we want to use int64 instead https://github.com/DIMO-Network/shared/pull/55

File: internal/services/elastic/elastic.go (hunk 29 above)
Comment (Allyson-English) [reply internal/services/elastic/elastic.go:157]: hmm I don't think so actually, looking at it now it seems like it'd make more sense to be moved up.. especially since minimum should make on just one item seems unnecessary

File: internal/services/elastic/elastic.go (hunk 9 above)
Comment (KevinJoiner) [reply internal/services/elastic/elastic.go:143]: Yeah I added these test cases https://github.com/DIMO-Network/device-data-api/pull/73/files#diff-5f7a333a7bff5d6ec59400aef69140e2ce90178144803dec0def9180d6379095R293-R314
One creates a large interval value, and the other creates a large number of buckets

**Review by elffjs — APPROVED** [review]

**Comment by elffjs** [comment]
Bah, we've got some linter failures on main don't we.

**Comment by elffjs** [comment]
I actually don't know what it's complaining about. Passes on my machine.

**Comment by elffjs** [comment]
I think we are good to merge this.

**Comment by KevinJoiner** [comment]
I had to update my local golangci-lint and then I was able to see all the linter errors.

**Comment by elffjs** [comment]
> I had to update my local golangci-lint and then I was able to see all the linter errors.

I thought I did that. Okay, one more time.

Oh! New version yesterday. Probably Homebrew hadn't picked it up yet. That's exciting.

**Comment by elffjs** [comment]
Fixed up on main just because.

**Review by elffjs — APPROVED** [review]
